### PR TITLE
remove legacy udf decorator

### PR DESCRIPTION
**Pull request by mattseddon**
This PR removes the legacy `@udf` decorator from the codebase. To do this I have migrated some tests from `DatasetQuery` to `DataChain`. 

I have deleted tests that relied on the udf decorator where: 

1. the code under test no longer seems relevant 
2. the test was explicitly testing the udf decorator
3. the code under test is covered elsewhere 
4. the test was marked xfail

### CODE REVIEW

**Review by rlamy — CHANGES_REQUESTED** [review]
Migrating all those tests is great! However, removing all tar and dir expansion tests is too much, we still need them!

**Comment by mattseddon** [comment]
> Migrating all those tests is great! However, removing all tar and dir expansion tests is too much, we still need them!

Why?

File: src/datachain/query/udf.py [hunk 1]
```diff
@@ -25,8 +24,6 @@
 )
 from .schema import (
     UDFParameter,
-    UDFParamSpec,
-    normalize_param,
 )
```
Comment (dreadatour) [inline src/datachain/query/udf.py:27]:
```python
from .schema import UDFParameter
```

File: tests/func/test_dataset_query.py [hunk 2]
```diff
@@ -1,63 +1,29 @@
...
-    DateTime,
-    Float,
-    Float32,
-    Float64,
-    Int,
-    Int32,
-    Int64,
-    SQLType,
     String,
 )
```
Comment (dreadatour) [inline tests/func/test_dataset_query.py:22]:
```python
from datachain.sql.types import String
```

File: tests/func/test_datasets.py [hunk 3]
```diff
@@ -12,20 +13,10 @@
...
-    Binary,
-    Boolean,
-    Float,
     Float32,
-    Float64,
-    Int,
-    Int32,
-    Int64,
-    String,
 )
```
Comment (dreadatour) [inline tests/func/test_datasets.py:20]:
```python
from datachain.sql.types import Float32
```

File: tests/func/test_dataset_query.py [hunk 4]
```diff
@@ -1,63 +1,29 @@
...
-from datachain.node import Node
 from datachain.query import (
     C,
     DatasetQuery,
-    DatasetRow,
-    LocalFilename,
     Object,
     Stream,
-    udf,
 )
```
Comment (dreadatour) [inline tests/func/test_dataset_query.py:17]:
```python
from datachain.query import C, DatasetQuery, Object, Stream
```

**Review by dreadatour — APPROVED** [review]
Looks good to me! Thank you! 👍

**Comment by rlamy** [comment]
> > Migrating all those tests is great! However, removing all tar and dir expansion tests is too much, we still need them!
> 
> Why?

* `test_dir_expansion()` is the only unit-test for that complicated piece of code that's critical for the Unix commands.
* The working tar tests are our only tests of tar file handling
* The xfailing tests for tar support in Unix commands are more debatable: we kept them because we want to support them again eventually.

**Comment by mattseddon** [comment]
> > > Migrating all those tests is great! However, removing all tar and dir expansion tests is too much, we still need them!
> > 
> > 
> > Why?
> 
> * `test_dir_expansion()` is the only unit-test for that complicated piece of code that's critical for the Unix commands.

It looks like the logic under test here is also tested by the following tests:

tests/func/test_catalog.py::test_cp_double_subdir
tests/func/test_catalog.py::test_cp_edatachain_file_options
tests/func/test_catalog.py::test_cp_multi_subdir
tests/func/test_catalog.py::test_cp_root
tests/func/test_catalog.py::test_cp_subdir
tests/func/test_catalog.py::test_cp_symlinks
tests/func/test_catalog.py::test_du
tests/func/test_catalog.py::test_find
tests/func/test_catalog.py::test_find_names_columns
tests/func/test_catalog.py::test_find_names_paths_size_type
tests/func/test_catalog.py::test_ls_glob
tests/func/test_ls.py::test_ls_glob_sub
tests/func/test_ls.py::test_ls_not_a_directory
tests/func/test_ls.py::test_ls_not_found
tests/func/test_ls.py::test_ls_partial_indexing
tests/func/test_ls.py::test_ls_sources
tests/func/test_ls.py::test_ls_sources_scheme_uppercased
tests/test_cli_e2e.py::test_cli_e2e
tests/test_query_e2e.py::test_query_e2e
tests/unit/test_listing.py::test_basic_expansion
tests/unit/test_listing.py::test_list_dir
tests/unit/test_listing.py::test_multilevel_expansion
tests/unit/test_listing.py::test_subdirs
tests/unit/test_listing.py::test_subname_expansion
tests/unit/test_listing.py::test_subtree

I have however updated `create_tar_dataset` to use the new generator function, renamed it to `create_tar_dataset_with_legacy_columns` and reinstated the `test_dir_expansion` test

> * The working tar tests are our only tests of tar file handling (Edit: see [Convert  `index_tar` to a new-style generator #439](https://github.com/iterative/datachain/issues/439))

I am assuming this is covered by the change you made in #440

> * The xfailing tests for tar support in Unix commands are more debatable: we kept them because we want to support them again eventually.

I am not putting any effort into bringing these tests back.

**Review by rlamy — COMMENTED** [review]
LGTM, except for the issue with `process_tar()`. Great clean-up!

**Comment by rlamy** [comment]
> I have however updated `create_tar_dataset` to use the new generator function, renamed it to `create_tar_dataset_with_legacy_columns` and reinstated the `test_dir_expansion` test

Thanks!
 
> > * The working tar tests are our only tests of tar file handling (Edit: see [Convert  `index_tar` to a new-style generator #439](https://github.com/iterative/datachain/issues/439))
> 
> I am assuming this is covered by the change you made in #440

Yes, it is.
 
> > * The xfailing tests for tar support in Unix commands are more debatable: we kept them because we want to support them again eventually.
> 
> I am not putting any effort into bringing these tests back.

That's fine, they're probably more trouble than they're worth.

File: src/datachain/lib/tar.py [hunk 5]
```diff
@@ -27,7 +27,9 @@ def build_tar_member(parent: File, info: tarfile.TarInfo) -> File:
 
 
 def process_tar(file: File) -> Iterator[File]:
+    yield file
     with file.open() as fd:
         with tarfile.open(fileobj=fd) as tar:
             for entry in tar.getmembers():
-                yield build_tar_member(file, entry)
+                if entry.isfile():
```
[the remark this answers is not in the document]
Comment (mattseddon) [reply src/datachain/lib/tar.py:34]: No, I caught this with `test_dir_expansion` but I've now extended `test_process_and_open_tar`.

File: src/datachain/lib/tar.py [hunk 6]
```diff
@@ -27,7 +27,9 @@ def build_tar_member(parent: File, info: tarfile.TarInfo) -> File:
 
 
 def process_tar(file: File) -> Iterator[File]:
+    yield file
```
[the remark this answers is not in the document]
Comment (mattseddon) [reply src/datachain/lib/tar.py]: removed and updated `create_tar_dataset_with_legacy_columns`

File: tests/unit/lib/test_datachain_merge.py [hunk 7]
```diff
@@ -293,3 +293,100 @@ def _get_expr(dc):
...
+
+    dogs1 = dogs.map(sig1=lambda: 1, output={"sig1": int})
+    dogs2 = dogs.map(sig2=lambda: 2, output={"sig2": int})
+    cats1 = cats.map(sig1=lambda: 1, output={"sig1": int})
+
+    joined = (dogs1 | cats1).merge(dogs2, "file.path", inner=inner)
+    signals = list(joined.order_by("file.path").collect("file.path", "sig1", "sig2"))
+
+    if cloud_type == "file":
+        prefix = cloud_test_catalog.partial_path + "/"
```
Comment (mattseddon) [inline tests/unit/lib/test_datachain_merge.py:318]: [I] make this into a fixture in a follow-up

**Pull request merged**